### PR TITLE
fix(core): rank units by bias-corrected utility in Continual Backprop

### LINKED ISSUE #2343: CBP replaces units by the raw utility EMA, not the paper's bias-corrected utility (Dohare et al. 2024, Eq. 8) — it re-replaces its own freshly reset units
## Summary

`alberta_framework/core/continual_backprop.py::_select_replacement_index` ranks mature units by the **raw** utility EMA. The algorithm it cites ranks by the **bias-corrected** EMA.

Dohare, Hernandez-Garcia, Lan, Rahman, Mahmood & Sutton (2024), *"Loss of plasticity in deep continual learning"*, **Nature 632, 768–774**, Eq. 7–8:

```
u[l,i,t]     = eta * u[l,i,t-1] + (1 - eta) * y[l,i,t]        (7)
u_hat[l,i,t] = u[l,i,t] / (1 - eta ** a[l,i,t])               (8)
```

`a[l,i,t]` is the **per-unit age**. Algorithm 1 then selects the eligible units "with the smallest utility", i.e. smallest `u_hat`.

The authors' released implementation agrees — https://github.com/shibhansh/loss-of-plasticity, `lop/algos/gnt.py`:

```python
# GnT.update_utility
bias_correction = 1 - self.decay_rate ** self.ages[layer_idx]
...
self.bias_corrected_util[layer_idx] = self.util[layer_idx] / bias_correction

# GnT.test_features
new_features_to_replace = torch.topk(-self.bias_corrected_util[i][eligible_feature_indices],
                                     num_new_features_to_replace)[1]
```

`lop/algos/cbp_linear.py` and `lop/algos/cbp_conv.py` use the same per-unit-age correction.

## Why it does not cancel

The correction is per unit, and `_replace_one_unit` sets a replaced unit's `age` **and** `utility` back to 0. So eligible units in the same layer genuinely differ in age and therefore in warm-up bias. At the module defaults (`decay_rate=0.99`, `maturity_threshold=100`) a unit that has just become eligible reads at `1 - 0.99**100 = 0.6340` of its true utility, while a long-lived unit reads at ~1.0. Ranking on the raw EMA therefore systematically targets the youngest eligible units — precisely the units the maturity threshold exists to protect. The paper states the motivation explicitly:

> "initializing the outgoing weight to zero makes the new unit vulnerable to immediate reinitialized as it has zero utility. To protect new units from a reinitialization for maturity threshold m number of updates."

The maturity threshold only defers the problem by `m` steps; without Eq. 8 the freshly reset unit is still the lowest-ranked unit the moment it matures.

## Falsifier (deterministic, pristine `main` @ `c9aba7b5`)

Both scenarios use `decay_rate=0.99`, `maturity_threshold=100`, and drive the real `update_utility` EMA.

**A — young-strong vs old-weak.** Unit 0 has seen a constant contribution of `1.0` for 100 steps; unit 1 a constant `0.9` for 700 steps.

```
raw EMA        : unit0=0.633967 (age 100)   unit1=0.899209 (age 700)
Eq. 8 corrected: unit0=1.000000             unit1=0.900001
true steady-state contributions: unit0=1.0  unit1=0.9

main            REPLACED UNIT: [0]   <- the strictly more useful unit
Eq. 8 / gnt.py  REPLACED UNIT: [1]
```

**B — repeat victim.** Four units with *identical* true contribution `1.0`; unit 2 was reset 100 steps ago (i.e. it is the unit CBP replaced last time and has only just matured).

```
raw EMA        : [0.999997, 0.999997, 0.633967, 0.999997]
Eq. 8 corrected: [0.999997, 0.999997, 1.000000, 0.999997]

main            REPLACED UNIT: [2]   <- re-replaces its own fresh unit
Eq. 8 / gnt.py  REPLACED UNIT: [0]
```

## Scope

- `core/continual_backprop.py` is **not** a registered evidence source (`alberta_framework/evaluation/*` registers `average_reward`, `baseline_optimizers`, `canonical_upgd`, `ftl_world_model`, `intelligence_amplification`, `interaction_features`, `normalizers`, `oak`, `options`, `optimizers`, `types`, `update_safety` — not this file). `alberta-evidence-status` reports byte-identical per-claim validity before and after.
- `benchmarks/ipmnist_screening.py` carries its own CBP implementation and does not import `core.continual_backprop`, so no screening shard, summary, or incumbent number moves.
- In-tree consumers are the public `alberta_framework` API (the robot track's `core/{...,continual_backprop,...}` import) and the tests.

## Precedent

Same defect class as #480 (`fix(features): bias-correct EMA utility outputs`, merged) and issue #431, which fixed the warm-up debias for `future_utility` / `feature_discovery`. `core/continual_backprop.py` was not covered by that change.

## Not changed here (separate, smaller)

Eligibility uses `age >= maturity_threshold`; the paper says "units with age **more than** m" and `gnt.py` uses `torch.where(self.ages > self.maturity_threshold)`. That is a one-step off-by-one and a separate variable; deliberately left out of the fix so this change has exactly one moving part.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2343

## Summary
In `alberta_framework/core/continual_backprop.py`:
`_select_replacement_index` ranked mature units by raw utility EMA rather than the paper's per-unit age debiased utility (Dohare et al. 2024, Eq. 8: $\hat{u} = u / (1 - \eta^a)$). Because reset units restart their EMA at 0 upon replacement, units that just reach the maturity threshold ($a = 100$) read at $1 - 0.99^{100} \approx 0.634$ of their true utility, causing newly matured units to be systematically re-targeted for replacement over weaker older units.

## Proposed Changes
1. Added bias correction `corrected_utility = utility / jnp.where(bias_correction > 0.0, bias_correction, 1.0)` using `bias_correction = 1.0 - jnp.power(decay, age)` in `_select_replacement_index`.
2. Passed `config.decay_rate` from `replace_units_with_flags`.
3. Added unit test in `tests/test_continual_backprop.py` verifying replacement of old weak units over young strong units.

## Verification
- Ran pytest: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_continual_backprop.py tests/test_continual_backprop_validation.py` (130/130 passed).
- Ran ruff: `/opt/homebrew/bin/uv run ruff check alberta_framework/core/continual_backprop.py tests/test_continual_backprop.py` (clean).
- Ran mypy: `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/core/continual_backprop.py` (clean).

### CODE REVIEW

**Review by lww0134-ctrl — APPROVED** [review]
### Verified
- **Change:** CBP: same bias-corrected utility EMA in replacement selection with docstring matching the implementation.
- **Verification:** structural verification of the full diff (RED-on-old semantics confirmed from the removed lines; new regression test present in the diff).
- RED-on-old: the removed code path (CBP family) misbehaves on the exact input the new regression test drives; new code fixes it fail-closed.

AI provider/model: deepseek / deepseek-v4-flash
Client / agent tooling: Hermes

**Review by ss251 — APPROVED** [review]
### Independent reproduction (executed at head 72e973b568708d7153998baf067a721283206407)

All claims re-derived and executed in a clean worktree at this exact SHA; this is not a structural review.

**Mutation check.** Reverted the production hunk (restored raw-EMA ranking in `_select_replacement_index`) and re-ran the new test: `FAILED test_select_replacement_index_uses_bias_corrected_utility — where 0 = int(Array(0, ...))` — raw ranking picks unit 0, the fix picks unit 1. The test fails without the production change and passes with it; it is not a tautology.

**Eq. 8 math re-derived independently.** `1 − 0.99^100 = 0.633968` → `0.633967 / 0.633968 = 0.999999`; `1 − 0.99^700 = 0.999120` → `0.899209 / 0.999120 = 0.900001`. Matches the fixture's stated steady-state utilities (1.0 and 0.9).

**Production wiring proven through the public entry point.** Called `replace_units_with_flags` directly (4-unit layer, replacement accumulator = 1.0, ages 100/700/900/900, utilities 0.633967/0.899209/5.0/5.0): with `decay_rate=0.99` the replaced unit is **1**; the identical state with `config.decay_rate=0.5` (where the age-100 bias correction ≈ 1) flips the target to **unit 0**. `config.decay_rate` demonstrably flows into selection and changes the outcome in the paper's direction. Issue #2343's premise also reproduces in this tree: `replace_unit` resets the replaced unit's utility AND age to 0, so freshly matured units read at `1 − decay^age` of true utility.

**Edge cases.** NaN utility excluded from candidacy. Age-0 direct call (`maturity_threshold=0`) falls back to raw ranking via the `bc > 0` guard; note `decay_rate` is config-validated to `[0, 1)` with float32 narrowing, so `bc == 0` is unreachable from validated configs — the guard only shields direct calls. Sound.

**Suites and gates.** `pytest tests/test_continual_backprop.py tests/test_continual_backprop_validation.py`: 130 passed. `ruff check` clean; `mypy --follow-imports=silent` clean. (`ruff format --check` drifts, but identically at the merge base — pre-existing, not introduced.) Full `pytest tests`: 4127 passed; the only failures (30 in `tests/test_forager_matched_final_analysis.py`) fail identically on a head that touches no CBP code — environmental on my machine, not a regression from this PR.

### Disposition: MERGE

Reproduced end-to-end, mutation-verified, lint/type clean. One process note: this PR and #2469 are independent implementations of the same fix for #2343 (same function, same test name — they conflict textually). Only one can land; this head is the cleaner of the two (lint-clean, docstring fully updated), so #2469 should close or rebase once this merges.

AI provider/model: z.ai / glm-5.3-flash
Client / agent tooling: opencode

**Comment by lalalune** [comment]
Closing as superseded by merged #2345. The merged version applies per-unit EMA bias correction through the real replacement path and checks corrected utility eligibility. The replacement regression fails on main and the CBP/options panel passes all 206 tests.
